<a href="https://colab.research.google.com/github/JordanHirsch/HealthcareAnalytics/blob/main/healthcare_data_generation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
from datetime import datetime, timedelta

In [ ]:
import os

BASE_DIR = "data/raw"
os.makedirs(BASE_DIR, exist_ok=True)

In [ ]:
departments = pd.DataFrame(columns=[
    "department_id",
    "department_name"
    "location",
    "budgeted_hours"
])

departments.to_csv(f"{BASE_DIR}/departments.csv", index=False)

In [ ]:
employees = pd.DataFrame(columns=[
    "employee_id",
    "department_id",
    "role",
    "hourly_rate",
    "hire_date",
    "employment_status"
])

employees.to_csv(f"{BASE_DIR}/empployees.csv", index=False)

In [ ]:
shifts = pd.DataFrame(columns=[
    "shift_id",
    "employee_id",
    "department_id",
    "shift_date",
    "schedules_hours",
    "worked_hours",
    "overtime_hours"
])

shifts.to_csv(f"{BASE_DIR}/shifts.csv", index=False)

In [ ]:
attendance = pd.DataFrame(columns=[
    "attendance_id",
    "employee_id",
    "shift_id",
    "status",
    "call_out_reason"
])

attendance.to_csv(f"{BASE_DIR}/attendance.csv", index=False)

In [ ]:
payroll = pd.DataFrame(columns=[
    "payroll_id",
    "employee_id",
    "pay_period",
    "regular_pay",
    "overtime_pay",
    "total_pay"
])

payroll.to_csv(f"{BASE_DIR}/payroll.csv", index=False)

In [ ]:
patient_survey = pd.DataFrame(columns=[
    "survey_id",
    "department_id",
    "visit_date",
    "satisfaction_score",
    "would_recommned"
])

patient_survey.to_csv(f"{BASE_DIR}/patient_surveys.csv", index=False)

In [ ]:
from google.colab import files

for file in os.listdir(BASE_DIR):
  files.download(f"{BASE_DIR}/{file}")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
departments = pd.DataFrame({
    "department_id": [1,2,3,4,5,6,7,8],
    "department_name": [
        "Emergency Department",
        "Med Surg",
        "ICU",
        "NICU",
        "Radiology",
        "Laboratory",
        "Pharmacy",
        "Registration"
    ],
    "department_type": [
        "Clinical",
        "Clinical",
        "Clinical",
        "Clinical",
        "Ancillary",
        "Ancillary",
        "Ancillary",
        "Administrative"
    ]
})

departments.to_csv("departments.csv", index=False)

In [ ]:
from google.colab import files

files.download("departments.csv")


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
import pandas as pd

employees = pd.DataFrame({
    "department_id": [
        1,1,1,1,          # Emergency
        2,2,2,            # Med Surg
        3,3,              # ICU
        4,4,              # NICU
        5,5,              # Radiology
        6,6,              # Laboratory
        7,7,              # Pharmacy
        8,8,8,8           # Registration
    ],
    "job_title": [
        "Registered Nurse","Registered Nurse","Nurse Technician","Charge Nurse",
        "Registered Nurse","Nurse Technician","Charge Nurse",
        "Registered Nurse","Charge Nurse",
        "Registered Nurse","Charge Nurse",
        "Radiology Technologist","Radiology Technologist",
        "Lab Technician","Lab Technician",
        "Pharmacist","Pharmacy Technician",
        "Registration Clerk","Registration Clerk","Patient Access Rep","Registration Supervisor"
    ],
    "employment_type": [
        "Full-Time","Full-Time","Part-Time","Full-Time",
        "Full-Time","Part-Time","Full-Time",
        "Full-Time","Full-Time",
        "Full-Time","Full-Time",
        "Full-Time","Part-Time",
        "Full-Time","Part-Time",
        "Full-Time","Part-Time",
        "Full-Time","Part-Time","Full-Time","Full-Time"
    ],
    "hourly_rate": [
        38,36,22,45,
        37,23,44,
        42,48,
        44,50,
        32,30,
        29,27,
        60,26,
        20,19,22,28
    ],
    "hire_date": pd.to_datetime([
        "2021-03-15","2022-07-10","2023-01-05","2020-11-01",
        "2021-06-20","2023-02-14","2019-09-30",
        "2020-08-18","2018-04-02",
        "2021-05-12","2017-10-22",
        "2022-01-19","2023-04-11",
        "2020-02-28","2022-12-05",
        "2019-06-07","2023-03-09",
        "2021-09-01","2023-06-15","2020-01-10","2018-07-25"
    ]),
    "status": ["Active"] * 21
})

# Create employee_id AFTER dataframe exists
employees.insert(0, "employee_id", range(1, len(employees) + 1))

employees.to_csv("employees.csv", index=False)
employees


employee_id  department_id                job_title employment_type  \
0             1              1         Registered Nurse       Full-Time   
1             2              1         Registered Nurse       Full-Time   
2             3              1         Nurse Technician       Part-Time   
3             4              1             Charge Nurse       Full-Time   
4             5              2         Registered Nurse       Full-Time   
5             6              2         Nurse Technician       Part-Time   
6             7              2             Charge Nurse       Full-Time   
7             8              3         Registered Nurse       Full-Time   
8             9              3             Charge Nurse       Full-Time   
9            10              4         Registered Nurse       Full-Time   
10           11              4             Charge Nurse       Full-Time   
11           12              5   Radiology Technologist       Full-Time   
12           13              5   Radiology Technologist       Part-Time   
13           14              6           Lab Technician       Full-Time   
14           15              6           Lab Technician       Part-Time   
15           16              7               Pharmacist       Full-Time   
16           17              7      Pharmacy Technician       Part-Time   
17           18              8       Registration Clerk       Full-Time   
18           19              8       Registration Clerk       Part-Time   
19           20              8       Patient Access Rep       Full-Time   
20           21              8  Registration Supervisor       Full-Time   

    hourly_rate  hire_date  status  
0            38 2021-03-15  Active  
1            36 2022-07-10  Active  
2            22 2023-01-05  Active  
3            45 2020-11-01  Active  
4            37 2021-06-20  Active  
5            23 2023-02-14  Active  
6            44 2019-09-30  Active  
7            42 2020-08-18  Active  
8            48 2018-04-02  Active  
9            44 2021-05-12  Active  
10           50 2017-10-22  Active  
11           32 2022-01-19  Active  
12           30 2023-04-11  Active  
13           29 2020-02-28  Active  
14           27 2022-12-05  Active  
15           60 2019-06-07  Active  
16           26 2023-03-09  Active  
17           20 2021-09-01  Active  
18           19 2023-06-15  Active  
19           22 2020-01-10  Active  
20           28 2018-07-25  Active

In [ ]:
from google.colab import files
files.download("employees.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
import pandas as pd
import numpy as np

np.random.seed(42)

num_shifts = 300  # hospital-scale sample
shift_dates = pd.date_range(start="2025-01-01", periods=90)

shifts = pd.DataFrame({
    "shift_id": range(1, num_shifts + 1),
    "employee_id": np.random.choice(range(1, 22), num_shifts),
    "department_id": np.random.choice(range(1, 9), num_shifts),
    "shift_date": np.random.choice(shift_dates, num_shifts),
    "scheduled_hours": 12,
    "actual_hours": np.random.choice([8, 10, 12, 14], num_shifts, p=[0.15, 0.25, 0.45, 0.15]),
    "shift_type": np.random.choice(["Day", "Night"], num_shifts, p=[0.6, 0.4])
})

# Calculate overtime
shifts["overtime_hours"] = shifts["actual_hours"].apply(lambda x: max(0, x - 12))

# Assign shift status
shifts["shift_status"] = np.where(shifts["actual_hours"] == 0, "Call-Out", "Completed")

shifts = shifts.sort_values("shift_date")

shifts.to_csv("shifts.csv", index=False)
shifts.head()


shift_id  employee_id  department_id shift_date  scheduled_hours  \
97         98           10              1 2025-01-01               12   
297       298           17              3 2025-01-01               12   
128       129            7              4 2025-01-01               12   
207       208           20              6 2025-01-01               12   
188       189           21              6 2025-01-01               12   

     actual_hours shift_type  overtime_hours shift_status  
97              8      Night               0    Completed  
297            12        Day               0    Completed  
128            10        Day               0    Completed  
207            14        Day               2    Completed  
188             8        Day               0    Completed

In [ ]:
from google.colab import files
files.download("shifts.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
np.random.seed(99)

num_records = 90 # Attendance exceptions
attendance_dates = pd.date_range(start="2024-01-01", periods=90)

attendance = pd.DataFrame({
     "attendance_id": range(1, num_records + 1),
    "employee_id": np.random.choice(range(1, 22), num_records),
    "department_id": np.random.choice(range(1, 9), num_records),
    "shift_date": np.random.choice(attendance_dates, num_records),
    "attendance_type": np.random.choice(
        ["Call-Out", "Late Arrival", "Left Early"],
        num_records,
        p=[0.45, 0.30, 0.25]
    ),
    "reason": np.random.choice(
        ["Sick", "Family Emergency", "Burnout", "Transportation", "Other"],
        num_records,
        p=[0.35, 0.20, 0.20, 0.15, 0.10]
    )
})
# Assign hours missed based on attendance type
attendance["hours_missed"] = attendance["attendance_type"].map({
    "Call-Out": 12,
    "Late Arrival": 2,
    "Left Early": 3
})

attendance = attendance.sort_values("shift_date")

attendance.to_csv("attendance.csv", index=False)
attendance.head()

attendance_id  employee_id  department_id shift_date attendance_type  \
27             28           15              3 2024-01-01    Late Arrival   
82             83            6              2 2024-01-01    Late Arrival   
79             80           13              6 2024-01-02        Call-Out   
7               8            6              7 2024-01-02    Late Arrival   
51             52           13              7 2024-01-02    Late Arrival   

              reason  hours_missed  
27  Family Emergency             2  
82              Sick             2  
79           Burnout            12  
7               Sick             2  
51  Family Emergency             2

In [ ]:
from google.colab import files
files.download("attendance.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
import pandas as pd
import numpy as np

np.random.seed(123)

num_records = 300
pay_dates = pd.date_range(start="2024-01-07", periods=12, freq="W")

payroll = pd.DataFrame({
    "payroll_id": range(1, num_records + 1),
    "employee_id": np.random.choice(range(1, 22), num_records),
    "department_id": np.random.choice(range(1, 9), num_records),
    "pay_date": np.random.choice(pay_dates, num_records),
    "regular_hours": np.random.choice([32, 36, 40], num_records, p=[0.2, 0.3, 0.5]),
    "overtime_hours": np.random.choice([0, 4, 8], num_records, p=[0.6, 0.25, 0.15]),
    "hourly_rate": np.random.choice(
        [20, 22, 26, 29, 32, 36, 38, 42, 45, 50, 60],
        num_records
    )
})

# Calculate pay
payroll["regular_pay"] = payroll["regular_hours"] * payroll["hourly_rate"]
payroll["overtime_pay"] = payroll["overtime_hours"] * payroll["hourly_rate"] * 1.5
payroll["total_pay"] = payroll["regular_pay"] + payroll["overtime_pay"]

payroll = payroll.sort_values("pay_date")

payroll.to_csv("payroll.csv", index=False)
payroll.head()


payroll_id  employee_id  department_id   pay_date  regular_hours  \
130         131            7              3 2024-01-07             40   
195         196            6              7 2024-01-07             40   
26           27            8              1 2024-01-07             32   
97           98            7              6 2024-01-07             36   
260         261            8              3 2024-01-07             40   

     overtime_hours  hourly_rate  regular_pay  overtime_pay  total_pay  
130               4           45         1800         270.0     2070.0  
195               0           36         1440           0.0     1440.0  
26                0           36         1152           0.0     1152.0  
97                0           42         1512           0.0     1512.0  
260               4           50         2000         300.0     2300.0

In [ ]:
from google.colab import files
files.download("payroll.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
import pandas as pd
import numpy as np

np.random.seed(202)

num_surveys = 200
survey_dates = pd.date_range(start="2024-01-01", periods=90)

patient_surveys = pd.DataFrame({
    "survey_id": range(1, num_surveys + 1),
    "department_id": np.random.choice(range(1, 9), num_surveys),
    "survey_date": np.random.choice(survey_dates, num_surveys),
    "satisfaction_score": np.random.choice(
        [1, 2, 3, 4, 5],
        num_surveys,
        p=[0.05, 0.10, 0.20, 0.35, 0.30]
    ),
    "would_recommend": np.random.choice(
        ["Yes", "No"],
        num_surveys,
        p=[0.82, 0.18]
    ),
    "wait_time_minutes": np.random.choice(
        [10, 20, 30, 45, 60, 90],
        num_surveys,
        p=[0.20, 0.25, 0.25, 0.15, 0.10, 0.05]
    ),
    "staffing_rating": np.random.choice(
        [1, 2, 3, 4, 5],
        num_surveys,
        p=[0.05, 0.10, 0.25, 0.35, 0.25]
    )
})

patient_surveys = patient_surveys.sort_values("survey_date")

patient_surveys.to_csv("patient_surveys.csv", index=False)
patient_surveys.head()


survey_id  department_id survey_date  satisfaction_score would_recommend  \
156        157              6  2024-01-01                   4             Yes   
136        137              6  2024-01-01                   5             Yes   
94          95              8  2024-01-01                   3             Yes   
53          54              1  2024-01-02                   2              No   
59          60              6  2024-01-02                   5             Yes   

     wait_time_minutes  staffing_rating  
156                 20                5  
136                 60                3  
94                  10                2  
53                  45                2  
59                  30                3

In [ ]:
from google.colab import files
files.download("patient_surveys.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
import sqlite3

# Create a SQLite database file
conn = sqlite3.connect("hospital_analytics.db")

# Create a cursor
cursor = conn.cursor()

print("SQLite database created and connected successfully.")

SQLite database created and connected successfully.


In [ ]:
# Confirm the DB exist
!ls

attendance.csv	departments.csv  hospital_analytics.db	payroll.csv  shifts.csv
data		employees.csv	 patient_surveys.csv	sample_data


In [ ]:
from google.colab import files

uploaded = files.upload()

KeyboardInterrupt: 

In [ ]:
# Check for files again
!ls

attendance.csv	departments.csv  hospital_analytics.db	payroll.csv  shifts.csv
data		employees.csv	 patient_surveys.csv	sample_data


In [ ]:
import pandas as pd
import sqlite3

# Connect to SQLite
conn = sqlite3.connect("hospital_analytics.db")

# List of CSVs and table names
tables = {
    "departments.csv": "departments",
    "employees.csv": "employees",
    "attendance.csv": "attendance"
}

for csv_file, table_name in tables.items():
    df = pd.read_csv(csv_file)
    df.to_sql(table_name, conn, if_exists="replace", index=False)
    print(f"{table_name} loaded successfully")

conn.close()


departments loaded successfully
employees loaded successfully
attendance loaded successfully


In [ ]:
# Test the connection and database
conn = sqlite3.connect("hospital_analytics.db")

pd.read_sql_query("SELECT * FROM employees LIMIT 5", conn)

employee_id  department_id         job_title employment_type  hourly_rate  \
0            1              1  Registered Nurse       Full-Time           38   
1            2              1  Registered Nurse       Full-Time           36   
2            3              1  Nurse Technician       Part-Time           22   
3            4              1      Charge Nurse       Full-Time           45   
4            5              2  Registered Nurse       Full-Time           37   

    hire_date  status  
0  2021-03-15  Active  
1  2022-07-10  Active  
2  2023-01-05  Active  
3  2020-11-01  Active  
4  2021-06-20  Active

In [ ]:
conn = sqlite3.connect("hospital_analytics.db")

query = """
SELECT
    d.department_name,
    COUNT(e.employee_id) AS employee_count
FROM employees e
JOIN departments d
    ON e.department_id = d.department_id
GROUP BY d.department_name
ORDER BY employee_count DESC;
"""

pd.read_sql_query(query, conn)

department_name  employee_count
0          Registration               4
1  Emergency Department               4
2              Med Surg               3
3             Radiology               2
4              Pharmacy               2
5                  NICU               2
6            Laboratory               2
7                   ICU               2

### Staffing Distribution Summary

This analysis shows how employees are distributed across departments within the facility.

Key Observations:
- Staffing levels vary significantly by department, reflecting differences in patient volume, acuity, and operational requirements.
- Higher headcount departments likely represent front-line or high-throughput areas (e.g., Emergency Department, Med Surge).
- Lower headcount departments may indicate specialized or support functions that require fewer staff but higher skill levels.

This metric establishes a baseline for evaluating labor costs, staffing efficiency, and reliability in subsequent analyses.

In [ ]:
query = """
SELECT
    d.department_name,
    ROUND(AVG(e.hourly_rate), 2) AS avg_hourly_rate
FROM employees e
JOIN departments d
    ON e.department_id = d.department_id
GROUP BY d.department_name
ORDER BY avg_hourly_rate DESC;
"""

pd.read_sql_query(query, conn)


department_name  avg_hourly_rate
0                  NICU            47.00
1                   ICU            45.00
2              Pharmacy            43.00
3  Emergency Department            35.25
4              Med Surg            34.67
5             Radiology            31.00
6            Laboratory            28.00
7          Registration            22.25

### Why this matters

This is important in understanding the headcount and labor cost correlations. Average hourly rate helps isolate whether cost differences are driving by staffing volume or workforce mix. This also shows that higher clinical acuity or specialized skills reflect higher average hourly rates.

This metric will be used alongside staffing levels and attendance data to evaluate labor efficiency and cost optimization opportunities.

In [ ]:
query = """
SELECT
    d.department_name,
    COUNT(a.shift_date) AS scheduled_shifts,
    SUM(CASE WHEN a.attendance_type = 'Call Out' THEN 1 ELSE 0 END) AS call_outs,
    ROUND(
        SUM(CASE WHEN a.attendance_type = 'Call Out' THEN 1 ELSE 0 END) * 1.0
        / COUNT(a.shift_date),
        3
    ) AS call_out_rate
FROM attendance a
JOIN employees e
    ON a.employee_id = e.employee_id
JOIN departments d
    ON e.department_id = d.department_id
GROUP BY d.department_name
ORDER BY call_out_rate DESC;
"""

pd.read_sql_query(query, conn)


department_name  scheduled_shifts  call_outs  call_out_rate
0          Registration                14          0            0.0
1             Radiology                15          0            0.0
2              Pharmacy                 6          0            0.0
3                  NICU                10          0            0.0
4              Med Surg                11          0            0.0
5            Laboratory                 5          0            0.0
6                   ICU                 9          0            0.0
7  Emergency Department                20          0            0.0

### Staffing Reliability by Department

This KPI measures employee call-out rates across departments, highlighting areas where staffing reliability may impact operations.

Key observations:
- Departments with higher call-out rates face increased risk of understaffing, overtime usage, and workflow disruption.
- Reliability challenges can compound labor costs and negatively affect patient experience.
- These results provide a foundation for targeted staffing strategies and proactive workforce planning.

This analysis links workforce behavior directly to operational risk and cost control.


In [ ]:
query = """
SELECT
    d.department_name,
    ROUND(SUM(a.worked_hours), 1) AS total_worked_hours,
    ROUND(SUM(a.scheduled_hours), 1) AS total_scheduled_hours,
    ROUND(
        SUM(a.worked_hours) * 1.0 / SUM(a.scheduled_hours),
        3
    ) AS utilization_rate
FROM attendance a
JOIN employees e
    ON a.employee_id = e.employee_id
JOIN departments d
    ON e.department_id = d.department_id
GROUP BY d.department_name
ORDER BY utilization_rate DESC;
"""

pd.read_sql_query(query, conn)


DatabaseError: Execution failed on sql '
SELECT
    d.department_name,
    ROUND(SUM(a.worked_hours), 1) AS total_worked_hours,
    ROUND(SUM(a.scheduled_hours), 1) AS total_scheduled_hours,
    ROUND(
        SUM(a.worked_hours) * 1.0 / SUM(a.scheduled_hours),
        3
    ) AS utilization_rate
FROM attendance a
JOIN employees e
    ON a.employee_id = e.employee_id
JOIN departments d
    ON e.department_id = d.department_id
GROUP BY d.department_name
ORDER BY utilization_rate DESC;
': no such column: a.worked_hours

In [ ]:
pd.read_sql_query("PRAGMA table_info(attendance);", conn)

cid             name     type  notnull dflt_value  pk
0    0    attendance_id  INTEGER        0       None   0
1    1      employee_id  INTEGER        0       None   0
2    2    department_id  INTEGER        0       None   0
3    3       shift_date     TEXT        0       None   0
4    4  attendance_type     TEXT        0       None   0
5    5           reason     TEXT        0       None   0
6    6     hours_missed  INTEGER        0       None   0

In [ ]:
query = """
SELECT
    d.department_name,
    COUNT(a.shift_date) AS scheduled_shifts,
    SUM(CASE WHEN a.attendance_type = 'Worked' THEN 1 ELSE 0 END) AS worked_shifts,
    ROUND(
        SUM(CASE WHEN a.attendance_type = 'Worked' THEN 1 ELSE 0 END) * 1.0
        / COUNT(a.shift_date),
        3
    ) AS shift_fulfillment_rate
FROM attendance a
JOIN employees e
    ON a.employee_id = e.employee_id
JOIN departments d
    ON e.department_id = d.department_id
GROUP BY d.department_name
ORDER BY shift_fulfillment_rate DESC;
"""

pd.read_sql_query(query, conn)


department_name  scheduled_shifts  worked_shifts  \
0          Registration                14              0   
1             Radiology                15              0   
2              Pharmacy                 6              0   
3                  NICU                10              0   
4              Med Surg                11              0   
5            Laboratory                 5              0   
6                   ICU                 9              0   
7  Emergency Department                20              0   

   shift_fulfillment_rate  
0                     0.0  
1                     0.0  
2                     0.0  
3                     0.0  
4                     0.0  
5                     0.0  
6                     0.0  
7                     0.0

In [ ]:
pd.read_sql_query("""
SELECT DISTINCT attendance_type
FROM attendance;
""", conn)

attendance_type
0    Late Arrival
1        Call-Out
2      Left Early

In [ ]:
pd.read_sql_query("""
SELECT
    d.department_name,
    COUNT(a.shift_date) AS scheduled_shifts,
    SUM(
        CASE
            WHEN a.attendance_type != 'Call-Out'
            THEN 1 ELSE 0
        END
    ) AS worked_shifts,
    ROUND(
        SUM(
            CASE
                WHEN a.attendance_type != 'Call-Out'
                THEN 1 ELSE 0
            END
        ) * 1.0 / COUNT(a.shift_date),
        3
    ) AS shift_fulfillment_rate
FROM attendance a
JOIN employees e
    ON a.employee_id = e.employee_id
JOIN departments d
    ON e.department_id = d.department_id
GROUP BY d.department_name
ORDER BY shift_fulfillment_rate DESC;
""", conn)


department_name  scheduled_shifts  worked_shifts  \
0  Emergency Department                20             17   
1              Med Surg                11              8   
2              Pharmacy                 6              4   
3             Radiology                15              9   
4                  NICU                10              6   
5            Laboratory                 5              3   
6          Registration                14              5   
7                   ICU                 9              2   

   shift_fulfillment_rate  
0                   0.850  
1                   0.727  
2                   0.667  
3                   0.600  
4                   0.600  
5                   0.600  
6                   0.357  
7                   0.222

### KPI Insight – Shift Fulfillment Rate
Departments with lower fulfillment rates may experience staffing gaps driven by call-outs.
This KPI helps identify units that may benefit from staffing buffers, cross-training, or schedule optimization.


## Does staffing reliability impact patient satisfaction?

### Business logic:
If call-outs, late arrivals, and early departures increase, patient satisfaction should drop.

In [57]:
query = """
SELECT
    d.department_name,
    COUNT(a.attendance_id) AS total_attendance_events,
    SUM(CASE WHEN a.attendance_type = 'Call-Out' THEN 1 ELSE 0 END) AS call_outs,
    ROUND(
        SUM(CASE WHEN a.attendance_type = 'Call-Out' THEN 1 ELSE 0 END) * 1.0
        / COUNT(a.attendance_id),
        3
    ) AS call_out_rate,
    ROUND(AVG(ps.satisfaction_score), 2) AS avg_patient_satisfaction
FROM attendance a
JOIN employees e
    ON a.employee_id = e.employee_id
JOIN departments d
    ON e.department_id = d.department_id
JOIN patient_surveys ps
    ON d.department_id = ps.department_id
GROUP BY d.department_name
ORDER BY call_out_rate DESC;
"""

pd.read_sql_query(query, conn)


department_name  total_attendance_events  call_outs  call_out_rate  \
0                   ICU                      234        182          0.778   
1          Registration                      392        252          0.643   
2             Radiology                      315        126          0.400   
3                  NICU                      230         92          0.400   
4            Laboratory                      150         60          0.400   
5              Pharmacy                      144         48          0.333   
6              Med Surg                      264         72          0.273   
7  Emergency Department                      480         72          0.150   

   avg_patient_satisfaction  
0                      4.00  
1                      4.18  
2                      3.71  
3                      4.22  
4                      4.13  
5                      3.00  
6                      3.75  
7                      3.75

### Staffing Reliability and Patient Satisfaction

This analysis examines the relationship between staffing reliability (measured by call-out rate) and average patient satisfaction across departments.

Key observations:
- The ICU and Registration departments exhibit the highest call-out rates (77.8% and 64.3% respectively), indicating frequent staffing disruptions.
- Departments with elevated call-out rates generally do not demonstrate the highest patient satisfaction scores, suggesting potential operational strain.
- The Emergency Department shows a notably lower call-out rate (15.0%) while maintaining a comparable patient satisfaction score, indicating more stable staffing coverage.
- Pharmacy and Med Surg display moderate to lower call-out rates but still show opportunities for improving patient satisfaction.

Overall, the results suggest a relationship between staffing reliability and patient experience. Departments experiencing frequent call-outs may benefit from targeted workforce interventions to stabilize coverage and protect service quality.


## Where are labor costs concentrated relative to staffing levels?

### Business logic:
High headcount + high pay = cost pressure.

In [58]:
query = """
SELECT
        d.department_name,
    COUNT(e.employee_id) AS employee_count,
    ROUND(AVG(e.hourly_rate), 2) AS avg_hourly_rate,
    ROUND(COUNT(e.employee_id) * AVG(e.hourly_rate), 2) AS staffing_cost_index
FROM employees e
JOIN departments d
    ON e.department_id = d.department_id
GROUP BY d.department_name
ORDER BY staffing_cost_index DESC;
"""
pd.read_sql_query(query, conn)


department_name  employee_count  avg_hourly_rate  staffing_cost_index
0  Emergency Department               4            35.25                141.0
1              Med Surg               3            34.67                104.0
2                  NICU               2            47.00                 94.0
3                   ICU               2            45.00                 90.0
4          Registration               4            22.25                 89.0
5              Pharmacy               2            43.00                 86.0
6             Radiology               2            31.00                 62.0
7            Laboratory               2            28.00                 56.0

### Labor Cost Concentration Across Departments

This analysis evaluates how staffing levels and average hourly pay combine to create labor cost pressure across departments.

Key observations:
- The Emergency Department represents the highest staffing cost concentration, driven by both higher headcount and elevated hourly rates.
- Med Surg also demonstrates significant cost exposure, suggesting staffing scale rather than pay rate alone contributes to labor costs.
- ICU and NICU show high average hourly rates but lower overall cost concentration due to smaller staff sizes.
- Registration maintains relatively high staffing levels but lower average pay, resulting in moderate cost pressure.
- Laboratory and Radiology exhibit the lowest staffing cost indices, indicating comparatively lower labor cost exposure.

These findings help identify where labor costs are most concentrated and where staffing adjustments may have the greatest financial impact.


## Which departments show optimization risk?

### Business logic:
Departments that are:
- High cost
- High reliability issues
- Lower patient satisfaction

are management risk zones.

In [59]:
query = """SELECT
    d.department_name,
    COUNT(DISTINCT e.employee_id) AS employee_count,
    ROUND(AVG(e.hourly_rate), 2) AS avg_hourly_rate,
    SUM(CASE WHEN a.attendance_type = 'Call-Out' THEN 1 ELSE 0 END) AS call_outs,
    ROUND(AVG(ps.satisfaction_score), 2) AS avg_patient_satisfaction
FROM departments d
LEFT JOIN employees e
    ON d.department_id = e.department_id
LEFT JOIN attendance a
    ON e.employee_id = a.employee_id
LEFT JOIN patient_surveys ps
    ON d.department_id = ps.department_id
GROUP BY d.department_name
ORDER BY call_outs DESC, avg_patient_satisfaction ASC;
"""
pd.read_sql_query(query, conn)


department_name  employee_count  avg_hourly_rate  call_outs  \
0          Registration               4            21.57        252   
1                   ICU               2            46.67        182   
2             Radiology               2            30.93        126   
3                  NICU               2            45.20         92   
4  Emergency Department               4            38.25         72   
5              Med Surg               3            31.91         72   
6            Laboratory               2            28.60         60   
7              Pharmacy               2            43.00         48   

   avg_patient_satisfaction  
0                      4.18  
1                      4.00  
2                      3.71  
3                      4.22  
4                      3.75  
5                      3.75  
6                      4.13  
7                      3.00

### Operational Risk Summary by Department

This analysis consolidates staffing levels, labor costs, workforce reliability, and patient satisfaction to identify departments with elevated operational risk.

Key observations:
- Registration and ICU experience the highest volume of call-outs, indicating significant staffing reliability challenges.
- Radiology and Pharmacy display lower patient satisfaction scores relative to other departments, suggesting potential service delivery pressures.
- ICU and NICU maintain higher average hourly rates, reflecting the specialized nature of care, but differ in workforce reliability and satisfaction outcomes.
- Emergency Department and Med Surg show moderate call-out volumes with average patient satisfaction, indicating potential opportunities for operational fine-tuning rather than major restructuring.

By evaluating these factors together, leadership can better identify departments where staffing interventions may improve performance without compromising care quality.
